packages/xstate-graph/src/graph.ts

```diff
@@ -393,64 +348,295 @@ export function toDirectedGraph(
-  let state = machine.initialState;
+  let state = behavior.initialState;
   for (const event of events) {
     path.push({
       state: stateMap.get(stateSerial)!,
       event
     });
 
-    const eventSerial = eventSerializer(event);
-    const nextSegment = adjacency[stateSerial][eventSerial];
+    const eventSerial = serializeEvent(event);
+    const { state: nextState, event: _nextEvent } = adjacency[
+      stateSerial
+    ].transitions[eventSerial];
 
-    if (!nextSegment) {
+    if (!nextState) {
       throw new Error(
         `Invalid transition from ${stateSerial} with ${eventSerial}`
       );
     }
 
-    const nextStateSerial = stateSerializer(nextSegment.state);
-    stateMap.set(nextStateSerial, nextSegment.state);
+    const nextStateSerial = serializeState(nextState, event);
+    stateMap.set(nextStateSerial, nextState);
 
     stateSerial = nextStateSerial;
-    state = nextSegment.state;
+    state = nextState;
   }
 
   return {
     state,
-    segments: path,
+    steps: path,
     weight: path.length
   };
 }
+
+interface AdjMap<TState, TEvent> {
+  [key: SerializedState]: {
+    state: TState;
+    transitions: {
+      [key: SerializedEvent]: {
+        event: TEvent;
+        state: TState;
+      };
+    };
+  };
+}
+
+export function performDepthFirstTraversal<TState, TEvent>(
+  behavior: SimpleBehavior<TState, TEvent>,
+  options: TraversalOptions<TState, TEvent>
+): AdjMap<TState, TEvent> {
+  const { transition, initialState } = behavior;
+  const {
+    serializeState,
+    getEvents,
+    traversalLimit: limit
+  } = resolveTraversalOptions(options);
+  const adj: AdjMap<TState, TEvent> = {};
+
+  let iterations = 0;
+  const queue: Array<[TState, TEvent | null]> = [[initialState, null]];
+
+  while (queue.length) {
+    const [state, event] = queue.shift()!;
+
+    if (iterations++ > limit) {
+      throw new Error('Traversal limit exceeded');
+    }
+
+    const serializedState = serializeState(state, event);
+    if (adj[serializedState]) {
+      continue;
+    }
+
+    adj[serializedState] = {
+      state,
+      transitions: {}
+    };
+
+    const events = getEvents(state);
+
+    for (const subEvent of events) {
+      const nextState = transition(state, subEvent);
+
+      if (!options.filter || options.filter(nextState, subEvent)) {
+        adj[serializedState].transitions[
+          JSON.stringify(subEvent) as SerializedEvent
+        ] = {
+          event: subEvent,
+          state: nextState
+        };
+        queue.push([nextState, subEvent]);
+      }
+    }
+  }
+
+  return adj;
+}
+
+function resolveTraversalOptions<TState, TEvent>(
+  traversalOptions?: Partial<TraversalOptions<TState, TEvent>>,
+  defaultOptions?: TraversalOptions<TState, TEvent>
+): Required<TraversalOptions<TState, TEvent>> {
+  const serializeState =
+    traversalOptions?.serializeState ??
+    defaultOptions?.serializeState ??
+    ((state) => JSON.stringify(state) as any);
+  return {
+    serializeState,
```

Comment:
I'm pretty sure that u can simplify this to this:
```suggestion
  return {
    serializeState: (state) => JSON.stringify(state) as any,
```
After all, you are already spreading `traversalOptions` and `defaultOptions` with the same "priorities" at the end of this object.